{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Normalize microarray expression for each of 6 donors to perform AUROC for SRP gene list individually on each donor\n",
    "- Get CHAT expression across sampled brain structures in each donor\n",
    "- Output values to csv file to generate plot in R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from pathlib import Path\n",
    "import os\n",
    "import pandas as pd\n",
    "from pandas.api.types import is_numeric_dtype\n",
    "from scipy.stats import zscore\n",
    "\n",
    "\n",
    "adult_data_path = Path('./data/raw/allen_HBA')\n",
    "fetal_data_path = Path('./data/raw/allen_human_fetal_brain')\n",
    "\n",
    "\n",
    "def read_expression_file(file_name):\n",
    "    expression_df = pd.read_csv(file_name, index_col=0, header=None)\n",
    "    expression_df.index.rename('probe_id', inplace=True)\n",
    "    return expression_df\n",
    "\n",
    "\n",
    "def read_samples_file(samples_file):\n",
    "    sample_df = pd.read_csv(samples_file)\n",
    "    sample_df.set_index(sample_df.index + 1, inplace=True)\n",
    "    sample_df.index.rename('sample_id', inplace=True)\n",
    "    return sample_df\n",
    "\n",
    "\n",
    "def get_probes_data(probes_file, probes_strategy='default'):\n",
    "    strats = ['default', 'reannotator', 'qc_filter', 'qc_scale']\n",
    "    assert probes_strategy in strats\n",
    "\n",
    "    # depending on strategy, may merge in diff tables to update probes info\n",
    "    probes_df = pd.read_csv(probes_file)\n",
    "\n",
    "    # rename columns for consistency between adult and fetal brain datasets\n",
    "    if 'probeset_name' in probes_df.columns:\n",
    "        probes_df.rename(columns={'probeset_name': 'probe_name',\n",
    "                                  'probeset_id': 'probe_id'}, inplace=True)\n",
    "    cols = ['probe_id', 'probe_name', 'gene_symbol']\n",
    "    probes_df = probes_df.loc[:, cols]\n",
    "\n",
    "    if probes_strategy == 'reannotator':\n",
    "        reannotations = get_probe_reannotations(\n",
    "            './data/raw/gene_symbol_annotations/AllenInstitute_custom_Agilent_Array.txt')\n",
    "        # drop the original gene_symbol column\n",
    "        probes_df.drop('gene_symbol', axis=1, inplace=True)\n",
    "        # merge in the reannotated gene_symbols\n",
    "        probes_df = probes_df.merge(reannotations, on='probe_name')\n",
    "\n",
    "    elif probes_strategy in ['qc_filter', 'qc_scale']:\n",
    "        # update gene_symbols using same method as was used in Rscript spatial transcriptomic analysis\n",
    "        probes_df.drop('gene_symbol', axis=1, inplace=True)\n",
    "        updated_annotations = pd.read_csv('./data/probe_annotations/updated_gene_symbol_annotations.csv')\n",
    "        probes_df = probes_df.merge(updated_annotations, on='probe_name', how='left')\n",
    "        #print(probes_df.head())\n",
    "        qc_filt = get_probe_qc_filter(\n",
    "            './data/probe_annotations/Miller et al. doi.org_10.1186_1471-2164-15-154 12864_2013_7016_MOESM8_ESM.xlsx')\n",
    "        probes_df = probes_df.merge(\n",
    "            qc_filt, left_on='probe_name', right_on='probe')\n",
    "        probes_df = probes_df[probes_df.qc_filter == True]\n",
    "        assert is_numeric_dtype(probes_df.m)\n",
    "        assert is_numeric_dtype(probes_df.b)\n",
    "\n",
    "        print('After getting probes_df which merged qc data, shape is {}'.format(\n",
    "            probes_df.shape))\n",
    "\n",
    "    probes_df.set_index('probe_id', inplace=True)\n",
    "\n",
    "    return probes_df\n",
    "\n",
    "\n",
    "def get_probe_reannotations(re_annotations_file):\n",
    "    # pre-processing function to prepare the probe reannotations file to be merge with probes df\n",
    "    re_annotations = pd.read_table(re_annotations_file,\n",
    "                                   usecols=['#PROBE_ID', 'Gene_symbol'])\n",
    "    re_annotations.rename(columns={'#PROBE_ID': 'probe_name'}, inplace=True)\n",
    "    re_annotations.dropna(inplace=True)\n",
    "    re_annotations.set_index('probe_name', inplace=True)\n",
    "    # split gene_symbols which have multiple genes associated with a single\n",
    "    # probe_name creates a new row for each of the gene_symbols\n",
    "    re_annotations = (re_annotations.Gene_symbol.str.split(';', expand=True)\n",
    "                                    .stack()\n",
    "                                    .reset_index()\n",
    "                                    .drop('level_1', axis=1)\n",
    "                                    .rename(columns={0: 'gene_symbol'}))\n",
    "\n",
    "    return re_annotations\n",
    "\n",
    "\n",
    "def get_probe_qc_filter(qc_file):\n",
    "    # pre-processes the Allen qc filter file\n",
    "    df = pd.read_excel(qc_file)\n",
    "    df = df.iloc[:, [0, 1, 2, 3, 7]].drop(0)  # .rename(columns=[col_names])\n",
    "    col_names = ['probe', 'gene_symbol', 'm', 'b', 'qc_filter']\n",
    "    df.columns = col_names\n",
    "    df[['m', 'b']] = df[['m', 'b']].apply(pd.to_numeric)\n",
    "    return df.drop('gene_symbol', axis=1)\n",
    "\n",
    "\n",
    "def scale_expression_vals(exp_df, probes_qc):\n",
    "    print('Expression df shape before scaling: {}'.format(exp_df.shape))\n",
    "    globScale = exp_df.values.flatten().mean()\n",
    "    exp_df = exp_df - globScale\n",
    "\n",
    "    df = exp_df.merge(probes_qc.loc[:, ['m', 'b']],\n",
    "                      left_index=True, right_index=True)\n",
    "    df.m = df.m.astype(float)\n",
    "    df.b = df.b.astype(float)\n",
    "    try:\n",
    "        df = df.apply(lambda x: df.m * x + df.b)\n",
    "    except Exception as e:\n",
    "        print('Exception occurred on row when applying scale fxn, skip row')\n",
    "        print(e)\n",
    "    df = df.drop(['m', 'b'], axis=1)\n",
    "\n",
    "    df.dropna(inplace=True)\n",
    "    print('Expression df shape after scaling: {}'.format(df.shape))\n",
    "    return df + globScale\n",
    "\n",
    "\n",
    "def get_donor_data(donor_file_list, probes_strategy):\n",
    "    # to work with both fetal and adult metadata\n",
    "    probe_file_strings = ['Probes', 'rows_meta']\n",
    "    samples_file_strings = ['Sample', 'columns_meta']\n",
    "    expression_file_strings = ['Expression', 'expression']\n",
    "\n",
    "    for file in donor_file_list:\n",
    "        if any(string in file.stem for string in probe_file_strings):\n",
    "            probes_df = get_probes_data(file, probes_strategy=probes_strategy)\n",
    "        elif any(string in file.stem for string in samples_file_strings):\n",
    "            samples_df = read_samples_file(file)\n",
    "        elif any(string in file.stem for string in expression_file_strings):\n",
    "            exp_df = read_expression_file(file)\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "    return exp_df, samples_df, probes_df\n",
    "\n",
    "\n",
    "def get_mean_expression_by_brain_area(exp_df, samples_df):\n",
    "    assert(exp_df.T.shape[0] == samples_df.shape[0])\n",
    "\n",
    "    # merge in metadata (brain area of sample)\n",
    "    annotated_df = exp_df.T.merge(samples_df[['structure_name']],\n",
    "                                  left_index=True, right_index=True)\n",
    "\n",
    "    # get mean expression level for samples within a brain area\n",
    "    expression_by_structure = annotated_df.groupby('structure_name').mean()\n",
    "    expression_by_structure.T.index.rename('gene_symbol', inplace=True)\n",
    "\n",
    "    return expression_by_structure.T\n",
    "\n",
    "\n",
    "def get_exp_by_genes(exp_df, probes_df):\n",
    "    \"\"\"\n",
    "    input is exp_df and probes metadata\n",
    "    output is exp_df grouped by gene_symbols and averaged\n",
    "    \"\"\"\n",
    "    annotated_df = exp_df.merge(probes_df[['gene_symbol']],\n",
    "                                left_index=True, right_index=True)\n",
    "\n",
    "    exp_by_genes = (annotated_df.groupby('gene_symbol')\n",
    "                                .mean())\n",
    "\n",
    "    return exp_by_genes\n",
    "\n",
    "\n",
    "def strip_left_right(structure_name):\n",
    "    brain_area_fragments = structure_name.split(',')\n",
    "    clean_fragments = []\n",
    "\n",
    "    for frag in brain_area_fragments:\n",
    "        if frag.strip() not in ['left', 'right', 'Left', 'Right']:\n",
    "            clean_fragments.append(frag)\n",
    "\n",
    "    clean_structure_name = ','.join(clean_fragments)\n",
    "    return clean_structure_name\n",
    "\n",
    "\n",
    "def get_single_donor_tidy_df(exp_df, samples_df, probes_df, donor_id, probes_strategy):\n",
    "    # remove left/right from brain structure_names\n",
    "    samples_df.structure_name = samples_df.structure_name.apply(\n",
    "        strip_left_right)\n",
    "\n",
    "    if probes_strategy == 'qc_scale':\n",
    "        exp_df = scale_expression_vals(exp_df, probes_df)\n",
    "        print('size of df after scaling: {}'.format(exp_df.shape))\n",
    "\n",
    "    # merge in probes metadata and get expression by gene_symbol\n",
    "    expression_by_genes = get_exp_by_genes(exp_df, probes_df)\n",
    "    print('size of df after merging probe info, grouping by gene: {}'.format(\n",
    "        expression_by_genes.shape))\n",
    "    \n",
    "    # merge in sample metadata and aggregate expression by brain area\n",
    "    exp_brain_area_by_genes = get_mean_expression_by_brain_area(expression_by_genes, samples_df)\n",
    "    \n",
    "\n",
    "    ranked_exp_by_area = exp_brain_area_by_genes.rank(ascending=True)\n",
    "    zscored_exp_by_area = pd.DataFrame(zscore(ranked_exp_by_area, axis=1), \n",
    "                                       index=ranked_exp_by_area.index,\n",
    "                                       columns= ranked_exp_by_area.columns)\n",
    "    #zscored_exp_by_area = ranked_exp_by_area.apply(zscore, axis=1, result_type='broadcast')\n",
    "\n",
    "    melted = pd.melt(zscored_exp_by_area.reset_index(),\n",
    "                     id_vars='gene_symbol',\n",
    "                     var_name='brain_area')\n",
    "    melted['donor_id'] = donor_id\n",
    "\n",
    "    return melted\n",
    "\n",
    "\n",
    "def generate_HBA_dataset(dataset, probes_strategy, selected_donor_ids=['10021', '9861', '14380', '15697', '15496', '12876']):\n",
    "    assert dataset in ['adult', 'fetal']\n",
    "    if dataset == 'adult':\n",
    "        print('--- Generating adult human HBA dataset ---')\n",
    "        #data_path = Path('./data/raw/allen_HBA')\n",
    "        data_path = adult_data_path\n",
    "    elif dataset == 'fetal':\n",
    "        print('--- Generating fetal human HBA dataset ---')\n",
    "        #data_path = Path('./data/raw/allen_human_fetal_brain')\n",
    "        data_path = fetal_data_path\n",
    "\n",
    "\n",
    "    selected_donor_paths = []\n",
    "    for donor_id in selected_donor_ids:\n",
    "        selected_donor_paths.append(list(data_path.glob(f'normalized_microarray*{donor_id}*')))\n",
    "\n",
    "    all_donors = []\n",
    "    for i, donor_folder in enumerate(selected_donor_paths):\n",
    "        print(f'Processing donor #{i+1}')\n",
    "        #donor_id = donor_folder.split('/')[-1].split('_')[-1]\n",
    "        #donor_folder = list(donor_folder)\n",
    "        donor_id = donor_folder[0].stem.split('_')[-1]\n",
    "        print(f'Donor ID: {donor_id}')\n",
    "        #donor_files = glob.glob(os.path.join(donor_folder, '*'))\n",
    "        donor_files = list(donor_folder[0].glob('*.csv'))\n",
    "        expression, samples, probes = get_donor_data(donor_files,\n",
    "                                                     probes_strategy)\n",
    "\n",
    "        tidy_donor = get_single_donor_tidy_df(expression,\n",
    "                                              samples,\n",
    "                                              probes,\n",
    "                                              donor_id=donor_id,\n",
    "                                              probes_strategy=probes_strategy)\n",
    "        all_donors.append(tidy_donor)\n",
    "\n",
    "    all_donors_long = pd.concat(all_donors)\n",
    "\n",
    "    structure_genes_exp_matrix = pd.pivot_table(all_donors_long,\n",
    "                                                values='value',\n",
    "                                                index='gene_symbol',\n",
    "                                                columns='brain_area')\n",
    "    return structure_genes_exp_matrix\n",
    "\n",
    "\n",
    "def get_dataset(dataset, probes_strategy, selected_donor_ids=['10021', '9861', '14380', '15697', '15496', '12876']):\n",
    "    # to reduce code duplication between get_HBA_dataset and\n",
    "    #  get_fetal_HBA_dataset\n",
    "    strats = ['reannotator', 'qc_filter', 'qc_scale', 'default']\n",
    "    datasets = ['adult', 'fetal']\n",
    "    assert probes_strategy in strats\n",
    "    assert dataset in datasets\n",
    "\n",
    "    donor_str = '-'.join(selected_donor_ids)\n",
    "    filename = '{0}_brainarea_vs_genes_exp_{1}_donors_{2}.tsv'.format(dataset,\n",
    "                                                                      probes_strategy, donor_str)\n",
    "    HBA_data_out_path = os.path.join('data', 'processed_HBA', filename)\n",
    "\n",
    "    if os.path.exists(HBA_data_out_path):\n",
    "        print(('Processed HBA brain dataset found locally. ' +\n",
    "               'Loading from {}'.format(HBA_data_out_path)))\n",
    "        structure_genes_exp_matrix = pd.read_csv(HBA_data_out_path,\n",
    "                                                 index_col='gene_symbol',\n",
    "                                                 sep='\\t')\n",
    "\n",
    "    else:\n",
    "        os.makedirs('./data/processed_HBA', exist_ok=True)\n",
    "        structure_genes_exp_matrix = generate_HBA_dataset(dataset,\n",
    "                                                          probes_strategy,\n",
    "                                                          selected_donor_ids=selected_donor_ids)\n",
    "        print('-- Writing data to ' + HBA_data_out_path + ' -- ')\n",
    "        structure_genes_exp_matrix.to_csv(HBA_data_out_path, sep='\\t')\n",
    "    return structure_genes_exp_matrix\n",
    "\n",
    "\n",
    "def generate_aggregate_data(probes_strategy='default', \n",
    "                             donor_ids=['9861','10021', '12876', '14380', '15496', '15697'], \n",
    "                             min_donors=1):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    data_path = adult_data_path\n",
    "    hba_donor_folders = [donor_dir for donor_dir in data_path.iterdir() if (donor_dir.is_dir() & ('microarray' in donor_dir.stem))]\n",
    "    \n",
    "    \n",
    "    selected_donor_paths = []\n",
    "    for path in hba_donor_folders:\n",
    "        if any(donor_id in path.stem for donor_id in donor_ids):\n",
    "            selected_donor_paths.append(path)\n",
    "    \n",
    "    all_donors = []\n",
    "    for i, donor_folder in enumerate(selected_donor_paths):\n",
    "        print(f'Processing donor #{i+1}')\n",
    "        print(f'Donor directory: {donor_folder}')\n",
    "        donor_id = donor_folder.stem\n",
    "        print(f'Donor ID: {donor_id}')\n",
    "        donor_files = list(donor_folder.glob('*')) \n",
    "        \n",
    "        expression, samples, probes = get_donor_data(donor_files, probes_strategy=probes_strategy)\n",
    "        donor_gene_exp = get_exp_by_genes(expression, probes)\n",
    "        samples.structure_name = samples.structure_name.apply(strip_left_right)\n",
    "        # this takes mean of gene expression from different brain structures\n",
    "        # an alternative way to try doing it is by concatenating all samples (from all donors) \n",
    "        # and average the final values\n",
    "        #exp_by_brain_area = merge_samples_w_exp(donor_gene_exp, samples)\n",
    "        exp_by_brain_area = get_mean_expression_by_brain_area(donor_gene_exp, samples)\n",
    "        #exp_by_brain_area['donor_id'] = donor_id\n",
    "        donor_expression  = exp_by_brain_area.T\n",
    "        donor_expression['donor_id'] = donor_id\n",
    "        \n",
    "        all_donors.append(donor_expression)\n",
    "        \n",
    "    concat_data = pd.concat(all_donors)\n",
    "    n_donors_per_structure = concat_data.groupby('structure_name').donor_id.nunique().reset_index()\n",
    "    selected_structures = n_donors_per_structure[n_donors_per_structure.donor_id >= min_donors].structure_name\n",
    "    # the index of concatenated data is the sampled structure_name\n",
    "    selected_samples = concat_data[concat_data.index.isin(selected_structures)]\n",
    "    final_matrix = selected_samples.groupby(['donor_id', 'structure_name']).mean().groupby('structure_name').mean()\n",
    "    \n",
    "    return final_matrix.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import mannwhitneyu\n",
    "import statsmodels.sandbox.stats.multicomp\n",
    "from sklearn import metrics\n",
    "\n",
    "\n",
    "def calc_pvalues(exp_series, gene_list):\n",
    "    X = exp_series[exp_series.index.isin(gene_list)]\n",
    "    y = exp_series[~exp_series.index.isin(gene_list)]\n",
    "    assert(X.shape[0] + y.shape[0] == exp_series.shape[0])\n",
    "\n",
    "    return mannwhitneyu(X, y, alternative='two-sided')[1]\n",
    "\n",
    "\n",
    "def calc_AUC(exp_series, gene_list):\n",
    "    # simplified function that calcs AUC for a single brain area series\n",
    "    y_true = exp_series.index.isin(gene_list)\n",
    "    y_score = exp_series.values\n",
    "\n",
    "    return metrics.roc_auc_score(y_true, y_score)\n",
    "\n",
    "\n",
    "def generate_stats_table(exp_df, gene_list, verbose=True):\n",
    "    \"\"\"\n",
    "    Creates a table of summary stats for each brain area\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    exp_df : dataframe\n",
    "        expression matrix: rows->genes ; columns->brain_areas\n",
    "    gene_list : series\n",
    "        list of gene symbols of interest\n",
    "    Returns\n",
    "    -------\n",
    "    table\n",
    "        results dataframe with brain areas as index\n",
    "\n",
    "    \"\"\"\n",
    "    count = len(gene_list)\n",
    "    n_genes_in_matrix = gene_list.isin(exp_df.index).sum()\n",
    "    genes_not_found = gene_list[~gene_list.isin(exp_df.index)].values\n",
    "\n",
    "    if verbose:\n",
    "        print('You submitted a gene list with {} genes.\\n\\\n",
    "    {} of those genes are present in the reference dataset.\\n\\\n",
    "    Genes not found in our reference data: {}'.format(\n",
    "            count, n_genes_in_matrix, genes_not_found))\n",
    "\n",
    "    pvalues = exp_df.apply(lambda col: calc_pvalues(\n",
    "        exp_series=col, gene_list=gene_list))\n",
    "    fdr_corrected = statsmodels.sandbox.stats.multicomp.multipletests(\n",
    "        pvalues, method=\"fdr_bh\")[1]\n",
    "    fdr_corrected = pd.Series(fdr_corrected, index=pvalues.index)\n",
    "    auc = exp_df.apply(lambda col: calc_AUC(\n",
    "        exp_series=col, gene_list=gene_list))\n",
    "\n",
    "    table = pd.concat([auc, pvalues, fdr_corrected],\n",
    "                      keys=['AUROC', 'p', 'pFDR', ], axis=1)\n",
    "    table.set_index(exp_df.columns, inplace=True)\n",
    "    return table.sort_values('AUROC', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process all 6 donors at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_10021-9861-14380-15697-15496-12876.tsv\n"
     ]
    }
   ],
   "source": [
    "all_donors = get_dataset(dataset='adult', probes_strategy='qc_filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "srp_genes = pd.read_csv('./data/gene_lists/GO.SRPdependentTranslationalProteinTargetingMembrane.txt', header=None, names=['gene_symbol'])\n",
    "srp_genes = srp_genes.loc[:, 'gene_symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n"
     ]
    }
   ],
   "source": [
    "all_donors_results = generate_stats_table(exp_df=all_donors, gene_list=srp_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AUROC</th>\n",
       "      <th>p</th>\n",
       "      <th>pFDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>substantia innominata</th>\n",
       "      <td>0.840875</td>\n",
       "      <td>4.668230e-28</td>\n",
       "      <td>2.166059e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleus accumbens</th>\n",
       "      <td>0.787368</td>\n",
       "      <td>2.080868e-20</td>\n",
       "      <td>3.713549e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>septal nuclei</th>\n",
       "      <td>0.784460</td>\n",
       "      <td>4.981247e-20</td>\n",
       "      <td>8.034375e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>central gray of the pons</th>\n",
       "      <td>0.777503</td>\n",
       "      <td>3.881691e-19</td>\n",
       "      <td>5.628452e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>medial parabrachial nucleus</th>\n",
       "      <td>0.773980</td>\n",
       "      <td>1.077122e-18</td>\n",
       "      <td>1.469955e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lateral parabrachial nucleus</th>\n",
       "      <td>0.773447</td>\n",
       "      <td>1.255559e-18</td>\n",
       "      <td>1.618276e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substantia nigra, pars reticulata</th>\n",
       "      <td>0.769991</td>\n",
       "      <td>3.369243e-18</td>\n",
       "      <td>3.908322e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, internal segment</th>\n",
       "      <td>0.765021</td>\n",
       "      <td>1.363611e-17</td>\n",
       "      <td>1.375469e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, external segment</th>\n",
       "      <td>0.764426</td>\n",
       "      <td>1.609533e-17</td>\n",
       "      <td>1.555882e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paraterminal gyrus</th>\n",
       "      <td>0.760793</td>\n",
       "      <td>4.391694e-17</td>\n",
       "      <td>3.918743e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      AUROC             p          pFDR\n",
       "substantia innominata              0.840875  4.668230e-28  2.166059e-26\n",
       "nucleus accumbens                  0.787368  2.080868e-20  3.713549e-19\n",
       "septal nuclei                      0.784460  4.981247e-20  8.034375e-19\n",
       "central gray of the pons           0.777503  3.881691e-19  5.628452e-18\n",
       "medial parabrachial nucleus        0.773980  1.077122e-18  1.469955e-17\n",
       "lateral parabrachial nucleus       0.773447  1.255559e-18  1.618276e-17\n",
       "substantia nigra, pars reticulata  0.769991  3.369243e-18  3.908322e-17\n",
       "globus pallidus, internal segment  0.765021  1.363611e-17  1.375469e-16\n",
       "globus pallidus, external segment  0.764426  1.609533e-17  1.555882e-16\n",
       "paraterminal gyrus                 0.760793  4.391694e-17  3.918743e-16"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_donors_results.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('./results/microarray/'):\n",
    "        os.makedirs('./results/microarray/')\n",
    "all_donors_results.to_csv('./results/microarray/srp_six_brains_regional_enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_14380-12876-15496-15697.tsv\n"
     ]
    }
   ],
   "source": [
    "four_donors = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['14380', '12876', '15496', '15697'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n"
     ]
    }
   ],
   "source": [
    "four_donors_results = generate_stats_table(exp_df=four_donors, gene_list=srp_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AUROC</th>\n",
       "      <th>p</th>\n",
       "      <th>pFDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>medial parabrachial nucleus</th>\n",
       "      <td>0.766002</td>\n",
       "      <td>1.037070e-17</td>\n",
       "      <td>1.273292e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subcallosal cingulate gyrus</th>\n",
       "      <td>0.759053</td>\n",
       "      <td>7.068262e-17</td>\n",
       "      <td>7.810430e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VI</th>\n",
       "      <td>0.753385</td>\n",
       "      <td>3.260264e-16</td>\n",
       "      <td>3.275084e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>head of caudate nucleus</th>\n",
       "      <td>0.752497</td>\n",
       "      <td>4.130169e-16</td>\n",
       "      <td>3.968553e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleus accumbens</th>\n",
       "      <td>0.750515</td>\n",
       "      <td>6.983407e-16</td>\n",
       "      <td>6.430554e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>septal nuclei</th>\n",
       "      <td>0.747778</td>\n",
       "      <td>1.432149e-15</td>\n",
       "      <td>1.266020e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substantia nigra, pars reticulata</th>\n",
       "      <td>0.745536</td>\n",
       "      <td>2.565192e-15</td>\n",
       "      <td>2.180413e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IV, lateral hemisphere</th>\n",
       "      <td>0.745119</td>\n",
       "      <td>2.857143e-15</td>\n",
       "      <td>2.338624e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, external segment</th>\n",
       "      <td>0.743500</td>\n",
       "      <td>4.335573e-15</td>\n",
       "      <td>3.304006e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, internal segment</th>\n",
       "      <td>0.738749</td>\n",
       "      <td>1.451444e-14</td>\n",
       "      <td>9.507221e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      AUROC             p          pFDR\n",
       "medial parabrachial nucleus        0.766002  1.037070e-17  1.273292e-16\n",
       "subcallosal cingulate gyrus        0.759053  7.068262e-17  7.810430e-16\n",
       "VI                                 0.753385  3.260264e-16  3.275084e-15\n",
       "head of caudate nucleus            0.752497  4.130169e-16  3.968553e-15\n",
       "nucleus accumbens                  0.750515  6.983407e-16  6.430554e-15\n",
       "septal nuclei                      0.747778  1.432149e-15  1.266020e-14\n",
       "substantia nigra, pars reticulata  0.745536  2.565192e-15  2.180413e-14\n",
       "IV, lateral hemisphere             0.745119  2.857143e-15  2.338624e-14\n",
       "globus pallidus, external segment  0.743500  4.335573e-15  3.304006e-14\n",
       "globus pallidus, internal segment  0.738749  1.451444e-14  9.507221e-14"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "four_donors_results.to_csv('./results/microarray/srp_four_brains_regional_enrichment.csv')\n",
    "four_donors_results.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process each donor separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_ids = ['10021', '9861', '14380', '15697', '15496', '12876']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_9861.tsv\n",
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_10021.tsv\n",
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_12876.tsv\n",
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_14380.tsv\n",
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_15496.tsv\n",
      "Processed HBA brain dataset found locally. Loading from data/processed_HBA/adult_brainarea_vs_genes_exp_qc_filter_donors_15697.tsv\n"
     ]
    }
   ],
   "source": [
    "donor_9861 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['9861'])\n",
    "donor_10021 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['10021'])\n",
    "donor_12876 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['12876'])\n",
    "donor_14380 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['14380'])\n",
    "donor_15496 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['15496'])\n",
    "donor_15697 = get_dataset(dataset='adult', probes_strategy='qc_filter', selected_donor_ids=['15697'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor9861\n",
      "Donor ID: normalized_microarray_donor9861\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n",
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor10021\n",
      "Donor ID: normalized_microarray_donor10021\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n",
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor12876\n",
      "Donor ID: normalized_microarray_donor12876\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n",
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor14380\n",
      "Donor ID: normalized_microarray_donor14380\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n",
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor15496\n",
      "Donor ID: normalized_microarray_donor15496\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n",
      "Processing donor #1\n",
      "Donor directory: data/raw/allen_HBA/normalized_microarray_donor15697\n",
      "Donor ID: normalized_microarray_donor15697\n",
      "After getting probes_df which merged qc data, shape is (31452, 8)\n"
     ]
    }
   ],
   "source": [
    "CHAT_9861 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['9861']).loc['CHAT']\n",
    "CHAT_10021 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['10021']).loc['CHAT']\n",
    "CHAT_12876 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['12876']).loc['CHAT']\n",
    "CHAT_14380 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['14380']).loc['CHAT']\n",
    "CHAT_15496 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['15496']).loc['CHAT']\n",
    "CHAT_15697 = generate_aggregate_data(probes_strategy='qc_filter', donor_ids=['15697']).loc['CHAT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "individual_donors = [donor_9861, donor_10021, donor_12876, donor_14380, donor_15496, donor_15697]\n",
    "CHAT_donors = [CHAT_9861, CHAT_10021, CHAT_12876, CHAT_14380, CHAT_15496, CHAT_15697]\n",
    "donor_ids = ['9861','10021', '12876', '14380', '15496', '15697']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n",
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n",
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n",
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n",
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n",
      "You submitted a gene list with 87 genes.\n",
      "    87 of those genes are present in the reference dataset.\n",
      "    Genes not found in our reference data: []\n"
     ]
    }
   ],
   "source": [
    "srp_results = {}\n",
    "chat_dfs = []\n",
    "\n",
    "for donor_id, donor, chat_exp in zip(donor_ids, individual_donors, CHAT_donors):\n",
    "    srp_results[donor_id] = generate_stats_table(exp_df=donor, gene_list=srp_genes)\n",
    "    srp_results[donor_id]['rank'] = srp_results[donor_id].AUROC.rank(ascending=False)\n",
    "    \n",
    "    chat_exp = chat_exp.reset_index()\n",
    "    chat_exp['donor_id'] = donor_id\n",
    "    chat_dfs.append(chat_exp)\n",
    "    \n",
    "srp_table = pd.concat(srp_results)\n",
    "chat_table = pd.concat(chat_dfs)\n",
    "\n",
    "srp_table = srp_table.rename_axis(['donor', 'brain_structure']).reset_index()\n",
    "srp_table = srp_table.rename(columns={'AUROC': 'srp_AUC'}).loc[:, ['brain_structure', 'srp_AUC', 'donor']]\n",
    "chat_table = chat_table.rename(columns={'structure_name': 'brain_structure', 'donor_id': 'donor'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 5 ranked brain structures for SRP enrichment for each donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>srp_AUC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor</th>\n",
       "      <th>brain_structure</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">10021</th>\n",
       "      <th>central gray of the pons</th>\n",
       "      <td>0.815393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleus accumbens</th>\n",
       "      <td>0.804307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amygdalohippocampal transition zone</th>\n",
       "      <td>0.791947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, internal segment</th>\n",
       "      <td>0.784864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zona incerta</th>\n",
       "      <td>0.777503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">12876</th>\n",
       "      <th>inferior temporal gyrus, lateral bank of gyrus</th>\n",
       "      <td>0.850401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subcallosal cingulate gyrus</th>\n",
       "      <td>0.843944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cingulate gyrus, frontal part, inferior bank of gyrus</th>\n",
       "      <td>0.817135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, external segment</th>\n",
       "      <td>0.791497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <td>0.761181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">14380</th>\n",
       "      <th>IV, lateral hemisphere</th>\n",
       "      <td>0.806822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>olfactory tubercle</th>\n",
       "      <td>0.773273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X, lateral hemisphere</th>\n",
       "      <td>0.766249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parahippocampal gyrus, lateral bank of gyrus</th>\n",
       "      <td>0.765835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, internal segment</th>\n",
       "      <td>0.761116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">15496</th>\n",
       "      <th>medial parabrachial nucleus</th>\n",
       "      <td>0.768014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substantia nigra, pars reticulata</th>\n",
       "      <td>0.749425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIIIB, paravermis</th>\n",
       "      <td>0.742377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, external segment</th>\n",
       "      <td>0.736923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>central nucleus</th>\n",
       "      <td>0.732746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">15697</th>\n",
       "      <th>VI</th>\n",
       "      <td>0.790434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>olfactory tubercle</th>\n",
       "      <td>0.789523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>III, paravermis</th>\n",
       "      <td>0.767711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>III, lateral hemisphere</th>\n",
       "      <td>0.767039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IV, lateral hemisphere</th>\n",
       "      <td>0.765660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">9861</th>\n",
       "      <th>subcallosal cingulate gyrus</th>\n",
       "      <td>0.827608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substantia innominata</th>\n",
       "      <td>0.819738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, internal segment</th>\n",
       "      <td>0.812479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>globus pallidus, external segment</th>\n",
       "      <td>0.797137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substantia nigra, pars reticulata</th>\n",
       "      <td>0.790316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           srp_AUC\n",
       "donor brain_structure                                             \n",
       "10021 central gray of the pons                            0.815393\n",
       "      nucleus accumbens                                   0.804307\n",
       "      amygdalohippocampal transition zone                 0.791947\n",
       "      globus pallidus, internal segment                   0.784864\n",
       "      zona incerta                                        0.777503\n",
       "12876 inferior temporal gyrus, lateral bank of gyrus      0.850401\n",
       "      subcallosal cingulate gyrus                         0.843944\n",
       "      cingulate gyrus, frontal part, inferior bank of...  0.817135\n",
       "      globus pallidus, external segment                   0.791497\n",
       "      temporal pole, medial aspect                        0.761181\n",
       "14380 IV, lateral hemisphere                              0.806822\n",
       "      olfactory tubercle                                  0.773273\n",
       "      X, lateral hemisphere                               0.766249\n",
       "      parahippocampal gyrus, lateral bank of gyrus        0.765835\n",
       "      globus pallidus, internal segment                   0.761116\n",
       "15496 medial parabrachial nucleus                         0.768014\n",
       "      substantia nigra, pars reticulata                   0.749425\n",
       "      VIIIB, paravermis                                   0.742377\n",
       "      globus pallidus, external segment                   0.736923\n",
       "      central nucleus                                     0.732746\n",
       "15697 VI                                                  0.790434\n",
       "      olfactory tubercle                                  0.789523\n",
       "      III, paravermis                                     0.767711\n",
       "      III, lateral hemisphere                             0.767039\n",
       "      IV, lateral hemisphere                              0.765660\n",
       "9861  subcallosal cingulate gyrus                         0.827608\n",
       "      substantia innominata                               0.819738\n",
       "      globus pallidus, internal segment                   0.812479\n",
       "      globus pallidus, external segment                   0.797137\n",
       "      substantia nigra, pars reticulata                   0.790316"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srp_table.set_index('brain_structure').groupby('donor').apply(lambda x: x.nlargest(5, 'srp_AUC')).drop('donor', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine SRP and CHAT enrichment data together and save as output for figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_table = chat_table.merge(srp_table, on=['brain_structure', 'donor'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_table.to_csv('./data/processed_HBA/srp_chat_enrichment_table.csv',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
